pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java

```diff
@@ -0,0 +1,300 @@
+        workerExecutor.shutdownNow();
+        performCheckpointExecutor.shutdownNow();
+        initiateCheckpointExecutor.shutdownNow();
+        readerGroup.close();
+        readerGroupManager.close();
+        writer.close();
+        clientFactory.close();
+        log.info("call: END");
+        return null;
+    }
+
+    /**
+     * Initiate a checkpoint, wait for it to complete, and write the checkpoint to the state.
+     */
+    private void performCheckpoint() {
+        final String checkpointName = UUID.randomUUID().toString();
+        log.info("performCheckpoint: BEGIN: checkpointName={}", checkpointName);
+        try {
+            Path checkpointDirPath = checkpointRootPath.resolve(checkpointName);
+            checkpointDirPath.toFile().mkdirs();
+
+            final Set<String> onlineReaders = readerGroup.getOnlineReaders();
+            log.info("performCheckpoint: onlineReaders ({})={}", onlineReaders.size(), onlineReaders);
+            log.info("performCheckpoint: Calling initiateCheckpoint; checkpointName={}", checkpointName);
+            CompletableFuture<Checkpoint> checkpointFuture = readerGroup.initiateCheckpoint(checkpointName, performCheckpointExecutor);
+            Checkpoint checkpoint = checkpointFuture.get(checkpointTimeoutMs, TimeUnit.MILLISECONDS);
+            // At this point, all workers have received and processed the checkpoint.
+            log.info("performCheckpoint: Checkpoint completed; checkpointName={}, positions={}", checkpointName, checkpoint.asImpl().getPositions());
+
+            Path checkpointPath = checkpointDirPath.resolve(PRAVEGA_CHECKPOINT_FILE_NAME);
+            log.info("Writing Pravega checkpoint to {}", checkpointPath);
+            try (FileOutputStream fos = new FileOutputStream(checkpointPath.toFile())) {
+                fos.write(checkpoint.toBytes().array());
+                fos.flush();
+                fos.getFD().sync();
+            }
+
+            // Create "latest" file that indicates the latest checkpoint name.
+            // This file must be updated atomically.
+            Path latestTmpCheckpointPath = checkpointRootPath.resolve(LATEST_CHECKPOINT_NAME_FILE_NAME + ".tmp");
+            try (FileOutputStream fos = new FileOutputStream(latestTmpCheckpointPath.toFile())) {
+                fos.write(checkpointName.getBytes(StandardCharsets.UTF_8));
+                fos.flush();
+                fos.getFD().sync();
+            }
+            Files.move(latestTmpCheckpointPath, latestCheckpointNamePath, StandardCopyOption.ATOMIC_MOVE);
+
+            // Read list of TxnIds from checkpoint directory written by all workers and commit all transactions.
+            commitTransactions(checkpointName);
+
+            cleanCheckpointDirectory(checkpointDirPath);
+        } catch (final Exception e) {
+            log.warn("performCheckpoint: Error performing checkpoint", e);
+            // Ignore error. We will retry when we are scheduled again.
```

Comment:
does it make sense to cleanup the checkpoint files in case of exception and commit any open transaction?

Reply:
If any exception occurs, then during the next execution of performCheckpoint, all previous transactions will be committed and old checkpoint files will be deleted. If this commits open transactions persisting the checkpoint to disk, then we may not have exactly once guarantees. The only problem I see is if exceptions occur continuously. In this case, new checkpoint directories will be created and never deleted. This app could put a limit on failures and exit if it exceeds a threshold. Do you think I should do this?

Reply:
Actually my comment above is incorrect. Transactions are associated with a single checkpoint. If that checkpoint fails, then all transactions associated with it need to be aborted and workers need to restart from the last successful checkpoint. The easiest way of doing this is to terminate and restart the application so the latest commit does this.

Reply:
I removed the exactly-once processor from this PR.